qiskit_optimization/algorithms/qrao/semideterministic_rounding.py

```diff
@@ -0,0 +1,80 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Semi-deterministic rounding module"""
+from __future__ import annotations
+
+import numpy as np
+
+from qiskit_optimization.algorithms import OptimizationResultStatus, SolutionSample
+from qiskit_optimization.exceptions import QiskitOptimizationError
+
+from .rounding_common import RoundingContext, RoundingResult, RoundingScheme
+
+
+class SemideterministicRounding(RoundingScheme):
+    """Semi-deterministic rounding scheme
+
+    This is referred to as "Pauli rounding" in
+    https://arxiv.org/abs/2111.03167v2.
+    """
+
+    def __init__(self, *, atol: float = 1e-8, seed: int | None = None):
+        """
+        Args:
+            seed: Seed for random number generator, which is used to resolve
+                expectation values near zero to either +1 or -1.
+            atol: Absolute tolerance for determining whether an expectation value is zero.
+        """
+        super().__init__()
+        self._rng = np.random.default_rng(seed)
+        self._atol = atol
+
+    def round(self, rounding_context: RoundingContext) -> RoundingResult:
+        """Perform semi-deterministic rounding
+
+        Args:
+            rounding_context: Rounding context containing information about the problem and solution.
+
+        Returns:
+            Result containing the rounded solution.
+
+        Raises:
+            QiskitOptimizationError: If the expectation values are not available in the context.
+        """
+        if rounding_context.expectation_values is None:
+            raise QiskitOptimizationError(
+                "Semideterministric rounding requires the expectation values of the ",
+                "``RoundingContext`` to be available, but they are not.",
+            )
+
+        rounded_vars = np.where(
+            np.isclose(rounding_context.expectation_values, 0, atol=self._atol),
+            self._rng.integers(2, size=len(rounding_context.expectation_values)),
+            np.less_equal(rounding_context.expectation_values, 0).astype(int),
```

Comment:
If expectation value of operators is close to 0, the corresponding binary variable value will be sampled randomly. So, there is an option to add `samples: int` argument to this function to generate solution samples.

Reply:
Here is an example of such case (symmetric problem instance). https://github.com/Qiskit/feedback/blob/main/demo-day-notebooks/2023-06-01/2_qrao_demoday230601.ipynb
The expectation values of operators corresponding to variables are all 0. So, the semi-deterministic rounding just randomly samples 0 or 1 for each variables.

Reply:
It might be useful if `SemidetterministicRounding` provides information about variables with expectation value 0. Users can know which variables were randomly sampled.